dali/operators/reader/loader/video/frames_decoder_test.cc

```diff
@@ -16,223 +16,219 @@
 
     // Seek to frame
     file.SeekFrame(25);
-    file.ReadNextFrame(frame);
-    copyD2H(frame_cpu.data(), frame.data(), frame.size());
-    this->CompareFramesAvgError(frame_cpu.data(), this->GetVfrFrame(1, 25), file.FrameSize());
+    ASSERT_EQ(file.NextFrameIdx(), 25);
+    file.ReadNextFrame(frame.data());
+    MemCopy(frame_cpu.data(), frame.data(), file.FrameSize());
+    ground_truth.CompareFrameAvgError(25, frame_cpu.data());
 
     // Seek back to frame
     file.SeekFrame(12);
-    file.ReadNextFrame(frame);
-    copyD2H(frame_cpu.data(), frame.data(), frame.size());
-    this->CompareFramesAvgError(frame_cpu.data(), this->GetVfrFrame(1, 12), file.FrameSize());
+    ASSERT_EQ(file.NextFrameIdx(), 12);
+    file.ReadNextFrame(frame.data());
+    MemCopy(frame_cpu.data(), frame.data(), file.FrameSize());
+    ground_truth.CompareFrameAvgError(12, frame_cpu.data());
 
     // Seek to last frame (flush frame)
-    file.SeekFrame(59);
-    file.ReadNextFrame(frame);
-    copyD2H(frame_cpu.data(), frame.data(), frame.size());
-    this->CompareFramesAvgError(frame_cpu.data(), this->GetVfrFrame(1, 59), file.FrameSize(), 1.1);
+    int last_frame_index = ground_truth.NumFrames() - 1;
+    file.SeekFrame(last_frame_index);
+    ASSERT_EQ(file.NextFrameIdx(), last_frame_index);
+    file.ReadNextFrame(frame.data());
+    MemCopy(frame_cpu.data(), frame.data(), file.FrameSize());
+    ground_truth.CompareFrameAvgError(last_frame_index, frame_cpu.data());
     ASSERT_EQ(file.NextFrameIdx(), -1);
 
     // Wrap around to first frame
-    ASSERT_FALSE(file.ReadNextFrame(frame));
+    ASSERT_FALSE(file.ReadNextFrame(frame.data()));
     file.Reset();
-    file.ReadNextFrame(frame);
-    copyD2H(frame_cpu.data(), frame.data(), frame.size());
-    this->CompareFramesAvgError(frame_cpu.data(), this->GetVfrFrame(1, 0), file.FrameSize());
+    ASSERT_EQ(file.NextFrameIdx(), 0);
+    file.ReadNextFrame(frame.data());
+    MemCopy(frame_cpu.data(), frame.data(), file.FrameSize());
+    ground_truth.CompareFrameAvgError(0, frame_cpu.data());
 
     // Seek to random frames and read them
     std::mt19937 gen(0);
-    std::uniform_int_distribution<> distr(0, file.NumFrames() - 1);
+    std::uniform_int_distribution<> distr(0, last_frame_index);
 
     for (int i = 0; i < 20; ++i) {
-        int next_index = distr(gen);
+      int next_index = distr(gen);
 
-        file.SeekFrame(next_index);
-        file.ReadNextFrame(frame);
-        copyD2H(frame_cpu.data(), frame.data(), frame.size());
-        this->CompareFramesAvgError(
-            frame_cpu.data(), this->GetVfrFrame(1, next_index), file.FrameSize());
+      file.SeekFrame(next_index);
+      file.ReadNextFrame(frame.data());
+      MemCopy(frame_cpu.data(), frame.data(), file.FrameSize());
+      ground_truth.CompareFrameAvgError(next_index, frame_cpu.data());
     }
+  }
+};
+
+
+TEST_F(FramesDecoderTest_CpuOnlyTests, ConstantFrameRate) {
+  RunTest("/db/video/cfr/test_1.mp4", cfr_videos_[0]);
```

Comment:
I wonder if we can have the path to the video and frames in the same place. Like:
```
RunTest("cfr_videos_[0].video",cfr_videos_[0].ref_frames);
```
?

Reply:
Done. Now I keep all paths in `VideoTestBase` in lists and match by index. I did to to be able to pass whole list to op specs in reader tests.